exchanges/kucoin/kucoin_test.go

```diff
@@ -0,0 +1,2562 @@
+	if err != nil {
+		t.Error("TransferFuturesFundsToMainAccount() error", err)
+	}
+}
+
+func TestTransferFundsToFuturesAccount(t *testing.T) {
+	t.Parallel()
+	sharedtestvalues.SkipTestIfCredentialsUnset(t, ku, canManipulateRealOrders)
+	if ku.CurrencyPairs.IsAssetEnabled(asset.Futures) != nil {
+		t.Skipf(assetNotEnabled, asset.Futures)
+	}
+	err := ku.TransferFundsToFuturesAccount(context.Background(), 1, "USDT", "MAIN")
+	if err != nil {
+		t.Error("TransferFundsToFuturesAccount() error", err)
+	}
+}
+
+func TestGetFuturesTransferOutList(t *testing.T) {
+	t.Parallel()
+	sharedtestvalues.SkipTestIfCredentialsUnset(t, ku)
+	if ku.CurrencyPairs.IsAssetEnabled(asset.Futures) != nil {
+		t.Skipf(assetNotEnabled, asset.Futures)
+	}
+	_, err := ku.GetFuturesTransferOutList(context.Background(), "USDT", "", time.Time{}, time.Time{})
+	if err != nil {
+		t.Error("GetFuturesTransferOutList() error", err)
+	}
+}
+
+func TestCancelFuturesTransferOut(t *testing.T) {
+	t.Parallel()
+	sharedtestvalues.SkipTestIfCredentialsUnset(t, ku, canManipulateRealOrders)
+	if ku.CurrencyPairs.IsAssetEnabled(asset.Futures) != nil {
+		t.Skipf(assetNotEnabled, asset.Futures)
+	}
+	err := ku.CancelFuturesTransferOut(context.Background(), "5cd53be30c19fc3754b60928")
+	if err != nil {
+		t.Error("CancelFuturesTransferOut() error", err)
+	}
+}
+
+func TestFetchTradablePairs(t *testing.T) {
+	t.Parallel()
+	_, err := ku.FetchTradablePairs(context.Background(), asset.Futures)
+	if err != nil {
+		t.Error(err)
+	}
+	_, err = ku.FetchTradablePairs(context.Background(), asset.Spot)
+	if err != nil {
+		t.Error(err)
+	}
+	_, err = ku.FetchTradablePairs(context.Background(), asset.Margin)
+	if err != nil {
+		t.Error(err)
+	}
+}
+
+func TestUpdateOrderbook(t *testing.T) {
+	t.Parallel()
+	if ku.CurrencyPairs.IsAssetEnabled(asset.Futures) == nil {
+		if _, err := ku.UpdateOrderbook(context.Background(), futuresTradablePair, asset.Futures); err != nil {
+			t.Error(err)
+		}
+	}
+	if ku.CurrencyPairs.IsAssetEnabled(asset.Margin) == nil {
+		if _, err := ku.UpdateOrderbook(context.Background(), marginTradablePair, asset.Margin); err != nil {
+			t.Error(err)
+		}
+	}
+	if ku.CurrencyPairs.IsAssetEnabled(asset.Spot) == nil {
+		if _, err := ku.UpdateOrderbook(context.Background(), spotTradablePair, asset.Spot); err != nil {
+			t.Error(err)
+		}
+	}
+}
+func TestUpdateTickers(t *testing.T) {
+	t.Parallel()
+	err := ku.UpdateTickers(context.Background(), asset.Spot)
+	if err != nil {
+		t.Fatal(err)
+	}
+	err = ku.UpdateTickers(context.Background(), asset.Margin)
+	if err != nil {
+		t.Fatal(err)
+	}
+	err = ku.UpdateTickers(context.Background(), asset.Futures)
+	if err != nil {
+		t.Fatal(err)
+	}
+	err = ku.UpdateTickers(context.Background(), asset.Empty)
+	if !errors.Is(err, asset.ErrNotSupported) {
+		t.Fatal(err)
+	}
+}
+func TestUpdateTicker(t *testing.T) {
+	t.Parallel()
+	var err error
+	if ku.CurrencyPairs.IsAssetEnabled(asset.Spot) == nil {
```

Comment:
The thing is I didn't use the asset type check for the first two endpoint calls in SubmitOrder because my aim was to validate the error response when the order side is not set and the Unsupported asset type is used.  And, I used the barrier because the credentials for Kucoin are not **unified**. We have to disable the futures asset type when we use SPOT-MARGIN credentials or disable SPOT-MARGIN assets when we use Futures credentials.

Reply:
If you don't think this reason is not sufficient to use the barriers, I can remove it.

Reply:
Its not really how it works in practice though. At present all assets are enabled and I can put in a spot or futures API key and all of the tests run anyway. If there was something that understood that you're running tests via futures or spot-margin and skipped tests appropriately it would make more sense to me.

So you can develop something to only run tests for the appropriate keys, but I don't think this is the solution for it

Also, this test is for updating the ticker, it does not require keys